{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymc3 as pm\n",
    "import pandas as pd\n",
    "from ray import tune\n",
    "\n",
    "#Se define función de entrenamiento que utilizó modelo bart_model para ajustar hiperparámetros en RayTune\n",
    "def train_bart(config):\n",
    "    num_trees = config[\"num_trees\"]\n",
    "    tree_depth = config[\"tree_depth\"]\n",
    "    alpha = config[\"alpha\"]\n",
    "\n",
    "    bart_model = bart_model(x, y, num_trees=num_trees, tree_depth=tree_depth, alpha=alpha)\n",
    "\n",
    "    with bart_model:\n",
    "        try:\n",
    "            trace = pm.sample(2000, tune=1000)\n",
    "            log_likelihood = -trace.log_likelihood.mean()\n",
    "        except Exception as e:\n",
    "            print(f\"Error during sampling: {e}\")\n",
    "            log_likelihood = float(\"-inf\")  # Valor predeterminado en caso de error\n",
    "\n",
    "#métrica a optimizar:en este caso, se utiliza el log-likelihood negativo como métrica para maximizar\n",
    "\n",
    "    return log_likelihood\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Definir espacio de búsqueda de hiperparámetros, para optimizar utilizando Ray Tune\n",
    "\n",
    "config = {\n",
    "    \"num_trees\": tune.choice([25, 50, 75]),\n",
    "    \"tree_depth\": tune.choice([2, 3, 4]),\n",
    "    \"alpha\": tune.choice([0.8, 0.9, 0.95]),\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-28 23:46:00,561\tINFO worker.py:1621 -- Started a local Ray instance.\n",
      "2023-07-28 23:46:02,326\tINFO tune.py:657 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"tuneStatus\">\n",
       "  <div style=\"display: flex;flex-direction: row\">\n",
       "    <div style=\"display: flex;flex-direction: column;\">\n",
       "      <h3>Tune Status</h3>\n",
       "      <table>\n",
       "<tbody>\n",
       "<tr><td>Current time:</td><td>2023-07-28 23:46:06</td></tr>\n",
       "<tr><td>Running for: </td><td>00:00:03.77        </td></tr>\n",
       "<tr><td>Memory:      </td><td>23.3/754.5 GiB     </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "    <div class=\"systemInfo\">\n",
       "      <h3>System Info</h3>\n",
       "      Using FIFO scheduling algorithm.<br>Logical resource usage: 3.0/96 CPUs, 0/0 GPUs\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "<div class=\"messages\">\n",
       "  <h3>Messages</h3>\n",
       "  \n",
       "  \n",
       "  Number of errored trials: 4<br><table>\n",
       "<thead>\n",
       "<tr><th>Trial name            </th><th style=\"text-align: right;\">  # failures</th><th>error file                                                                                                                                            </th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_bart_70235_00000</td><td style=\"text-align: right;\">           1</td><td>/home/info335/ray_results/train_bart_2023-07-28_23-46-02/train_bart_70235_00000_0_alpha=0.8000,num_trees=25,tree_depth=4_2023-07-28_23-46-02/error.txt</td></tr>\n",
       "<tr><td>train_bart_70235_00001</td><td style=\"text-align: right;\">           1</td><td>/home/info335/ray_results/train_bart_2023-07-28_23-46-02/train_bart_70235_00001_1_alpha=0.9000,num_trees=25,tree_depth=4_2023-07-28_23-46-02/error.txt</td></tr>\n",
       "<tr><td>train_bart_70235_00002</td><td style=\"text-align: right;\">           1</td><td>/home/info335/ray_results/train_bart_2023-07-28_23-46-02/train_bart_70235_00002_2_alpha=0.9500,num_trees=50,tree_depth=2_2023-07-28_23-46-02/error.txt</td></tr>\n",
       "<tr><td>train_bart_70235_00003</td><td style=\"text-align: right;\">           1</td><td>/home/info335/ray_results/train_bart_2023-07-28_23-46-02/train_bart_70235_00003_3_alpha=0.9500,num_trees=50,tree_depth=3_2023-07-28_23-46-02/error.txt</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>\n",
       "<style>\n",
       ".messages {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  padding-left: 1em;\n",
       "  overflow-y: auto;\n",
       "}\n",
       ".messages h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n",
       "\n",
       "  </div>\n",
       "  <div class=\"hDivider\"></div>\n",
       "  <div class=\"trialStatus\">\n",
       "    <h3>Trial Status</h3>\n",
       "    <table>\n",
       "<thead>\n",
       "<tr><th>Trial name            </th><th>status  </th><th>loc                  </th><th style=\"text-align: right;\">  alpha</th><th style=\"text-align: right;\">  num_trees</th><th style=\"text-align: right;\">  tree_depth</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_bart_70235_00000</td><td>ERROR   </td><td>146.83.216.133:999699</td><td style=\"text-align: right;\">   0.8 </td><td style=\"text-align: right;\">         25</td><td style=\"text-align: right;\">           4</td></tr>\n",
       "<tr><td>train_bart_70235_00001</td><td>ERROR   </td><td>146.83.216.133:999698</td><td style=\"text-align: right;\">   0.9 </td><td style=\"text-align: right;\">         25</td><td style=\"text-align: right;\">           4</td></tr>\n",
       "<tr><td>train_bart_70235_00002</td><td>ERROR   </td><td>146.83.216.133:999700</td><td style=\"text-align: right;\">   0.95</td><td style=\"text-align: right;\">         50</td><td style=\"text-align: right;\">           2</td></tr>\n",
       "<tr><td>train_bart_70235_00003</td><td>ERROR   </td><td>146.83.216.133:999701</td><td style=\"text-align: right;\">   0.95</td><td style=\"text-align: right;\">         50</td><td style=\"text-align: right;\">           3</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "  </div>\n",
       "</div>\n",
       "<style>\n",
       ".tuneStatus {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".tuneStatus .systemInfo {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       ".tuneStatus .trialStatus {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".tuneStatus .hDivider {\n",
       "  border-bottom-width: var(--jp-border-width);\n",
       "  border-bottom-color: var(--jp-border-color0);\n",
       "  border-bottom-style: solid;\n",
       "}\n",
       ".tuneStatus .vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=999699)\u001b[0m WARNING (theano.tensor.blas): Using NumPy C-API based implementation for BLAS functions.\n",
      "2023-07-28 23:46:06,085\tERROR tune_controller.py:911 -- Trial task failed for trial train_bart_70235_00000\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/worker.py\", line 2493, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(UnboundLocalError): \u001b[36mray::ImplicitFunc.train()\u001b[39m (pid=999699, ip=146.83.216.133, actor_id=03d9a9d256e8f902d19f3c1b01000000, repr=train_bart)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/trainable.py\", line 375, in train\n",
      "    raise skipped from exception_cause(skipped)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 349, in entrypoint\n",
      "    return self._trainable_func(\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 666, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"/tmp/ipykernel_954268/622252481.py\", line 11, in train_bart\n",
      "UnboundLocalError: local variable 'bart_model' referenced before assignment\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"trialProgress\">\n",
       "  <h3>Trial Progress</h3>\n",
       "  <table>\n",
       "<thead>\n",
       "<tr><th>Trial name            </th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_bart_70235_00000</td></tr>\n",
       "<tr><td>train_bart_70235_00001</td></tr>\n",
       "<tr><td>train_bart_70235_00002</td></tr>\n",
       "<tr><td>train_bart_70235_00003</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>\n",
       "<style>\n",
       ".trialProgress {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".trialProgress h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".trialProgress td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-28 23:46:06,091\tERROR tune_controller.py:911 -- Trial task failed for trial train_bart_70235_00001\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/worker.py\", line 2493, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(UnboundLocalError): \u001b[36mray::ImplicitFunc.train()\u001b[39m (pid=999698, ip=146.83.216.133, actor_id=04ec5f2640d576477adb76ad01000000, repr=train_bart)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/trainable.py\", line 375, in train\n",
      "    raise skipped from exception_cause(skipped)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 349, in entrypoint\n",
      "    return self._trainable_func(\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 666, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"/tmp/ipykernel_954268/622252481.py\", line 11, in train_bart\n",
      "UnboundLocalError: local variable 'bart_model' referenced before assignment\n",
      "2023-07-28 23:46:06,097\tERROR tune_controller.py:911 -- Trial task failed for trial train_bart_70235_00003\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/worker.py\", line 2493, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(UnboundLocalError): \u001b[36mray::ImplicitFunc.train()\u001b[39m (pid=999701, ip=146.83.216.133, actor_id=e7ba61c676071ac6d7dd791901000000, repr=train_bart)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/trainable.py\", line 375, in train\n",
      "    raise skipped from exception_cause(skipped)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 349, in entrypoint\n",
      "    return self._trainable_func(\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 666, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"/tmp/ipykernel_954268/622252481.py\", line 11, in train_bart\n",
      "UnboundLocalError: local variable 'bart_model' referenced before assignment\n",
      "2023-07-28 23:46:06,103\tERROR tune_controller.py:911 -- Trial task failed for trial train_bart_70235_00002\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/_private/worker.py\", line 2493, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(UnboundLocalError): \u001b[36mray::ImplicitFunc.train()\u001b[39m (pid=999700, ip=146.83.216.133, actor_id=f29a1516ccec1a297f76bc0701000000, repr=train_bart)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/trainable.py\", line 375, in train\n",
      "    raise skipped from exception_cause(skipped)\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 349, in entrypoint\n",
      "    return self._trainable_func(\n",
      "  File \"/home/info335/.local/lib/python3.10/site-packages/ray/tune/trainable/function_trainable.py\", line 666, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"/tmp/ipykernel_954268/622252481.py\", line 11, in train_bart\n",
      "UnboundLocalError: local variable 'bart_model' referenced before assignment\n"
     ]
    },
    {
     "ename": "TuneError",
     "evalue": "('Trials did not complete', [train_bart_70235_00000, train_bart_70235_00001, train_bart_70235_00002, train_bart_70235_00003])",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTuneError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 14\u001b[0m\n\u001b[1;32m     11\u001b[0m ray\u001b[39m.\u001b[39minit()\n\u001b[1;32m     13\u001b[0m \u001b[39m# Ejecución Ray Tune\u001b[39;00m\n\u001b[0;32m---> 14\u001b[0m analysis \u001b[39m=\u001b[39m tune\u001b[39m.\u001b[39;49mrun(\n\u001b[1;32m     15\u001b[0m     train_bart,\n\u001b[1;32m     16\u001b[0m     config\u001b[39m=\u001b[39;49mconfig,\n\u001b[1;32m     17\u001b[0m     metric\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mlog_likelihood\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m     18\u001b[0m     mode\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mmax\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m     19\u001b[0m     num_samples\u001b[39m=\u001b[39;49m\u001b[39m4\u001b[39;49m,\n\u001b[1;32m     20\u001b[0m     resources_per_trial\u001b[39m=\u001b[39;49m{\u001b[39m\"\u001b[39;49m\u001b[39mcpu\u001b[39;49m\u001b[39m\"\u001b[39;49m: \u001b[39m1\u001b[39;49m},\n\u001b[1;32m     21\u001b[0m     )\n\u001b[1;32m     23\u001b[0m \u001b[39m# Obtener los mejores hiperparámetros\u001b[39;00m\n\u001b[1;32m     24\u001b[0m best_config \u001b[39m=\u001b[39m analysis\u001b[39m.\u001b[39mget_best_config(metric\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mlog_likelihood\u001b[39m\u001b[39m\"\u001b[39m, mode\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mmax\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/ray/tune/tune.py:1142\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(run_or_experiment, name, metric, mode, stop, time_budget_s, config, resources_per_trial, num_samples, storage_path, search_alg, scheduler, checkpoint_config, verbose, progress_reporter, log_to_file, trial_name_creator, trial_dirname_creator, chdir_to_trial_dir, sync_config, export_formats, max_failures, fail_fast, restore, server_port, resume, reuse_actors, raise_on_failed_trial, callbacks, max_concurrent_trials, keep_checkpoints_num, checkpoint_score_attr, checkpoint_freq, checkpoint_at_end, checkpoint_keep_all_ranks, checkpoint_upload_from_workers, trial_executor, local_dir, _experiment_checkpoint_dir, _remote, _remote_string_queue, _entrypoint)\u001b[0m\n\u001b[1;32m   1140\u001b[0m \u001b[39mif\u001b[39;00m incomplete_trials:\n\u001b[1;32m   1141\u001b[0m     \u001b[39mif\u001b[39;00m raise_on_failed_trial \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m experiment_interrupted_event\u001b[39m.\u001b[39mis_set():\n\u001b[0;32m-> 1142\u001b[0m         \u001b[39mraise\u001b[39;00m TuneError(\u001b[39m\"\u001b[39m\u001b[39mTrials did not complete\u001b[39m\u001b[39m\"\u001b[39m, incomplete_trials)\n\u001b[1;32m   1143\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1144\u001b[0m         logger\u001b[39m.\u001b[39merror(\u001b[39m\"\u001b[39m\u001b[39mTrials did not complete: \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m, incomplete_trials)\n",
      "\u001b[0;31mTuneError\u001b[0m: ('Trials did not complete', [train_bart_70235_00000, train_bart_70235_00001, train_bart_70235_00002, train_bart_70235_00003])"
     ]
    }
   ],
   "source": [
    "#Ejecución de Ray Tune para optimizar hiperparámetros (utilizando función entrenamiento y espacio de búsqueda)\n",
    "import ray\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    data = pd.read_csv('Datos_Bart_C.csv', delimiter=';')\n",
    "    x = data[['total_credits_1', 'total_credits_2', 'total_courses_1', 'course_approved_1', 't_gpa_1']].values\n",
    "    y = data['Permanencia_term_3'].values\n",
    "\n",
    "    ray.shutdown()\n",
    "    # Inicia Ray \n",
    "    ray.init()\n",
    "\n",
    "    # Ejecución Ray Tune\n",
    "    analysis = tune.run(\n",
    "        train_bart,\n",
    "        config=config,\n",
    "        metric=\"log_likelihood\",\n",
    "        mode=\"max\",\n",
    "        num_samples=10,\n",
    "        resources_per_trial={\"cpu\": 1},\n",
    "        )\n",
    "\n",
    "    # Obtener los mejores hiperparámetros\n",
    "    best_config = analysis.get_best_config(metric=\"log_likelihood\", mode=\"max\")\n",
    "    print(\"Mejores hiperparámetros:\", best_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Forma de poder evaluar mejor modelo entrenado\n",
    "import os\n",
    "\n",
    "logdir = results.get_best_result(\"mean_accuracy\", mode=\"max\").log_dir\n",
    "state_dict = torch.load(os.path.join(logdir, \"model.pth\"))\n",
    "\n",
    "model = ConvNet()\n",
    "model.load_state_dict(state_dict)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
